README

 libnuosc++ - A library for calculating 3 neutrino oscillation probabilities.

Installation:
=============

See the INSTALL file.

Introduction:
=============

This library provides various methods to calculate 3 neutrino
oscillation probabilities with and with out matter effects as well as
the possibility for a nonzero Dirac CP phase.  Anti neutrinos are also
handled.

It does this by solving the evolution equation:

	d(nu)/dx = [-i/(2E) U M^2 Udagger -iV] nu(x)

where U is the (potentially complex) mixing matrix, M^2 is
diag(0,dm^2_21,dm^2_31) and V = diag(A,0,0) is the matter effect term.
In the case of anti neutrinos the sign of A and of the CP phase is
inverted.

There are three classes of calculations with regards to matter
effects: vacuum oscillations in the absence of matter, oscillations
through constant density matter and a spherically symmetric earth
matter density profile from the Preliminary Reference Earth Model
(PREM) (see hep-ph/0002149 and http://solid_earth.ou.edu/prem.html).

For each of these three classes there are two methods of calculation.
A direct matrix transform method and a stepped method.  The matrix
transform involves analytically solving the evolution equation (see
hep-ph/9910546).  This solution is exact for the case of constant
density (and of course vacuum).

The second method is to step the evolution equation down the neutrino
baseline.  This is done with an 5th order adaptive Runge-Kutta
stepper.  Stepping in inherently non-exact, but the precision can be
controlled to arbitrary tightness at the cost of running time.  For
the vacuum or constant density cases the matrix methods are obviously
preferable.  In these cases, the step methods were developed for
checking, debugging and building up to the last class of calculation.

Since the earth density profile is not constant there is no analytical
solution so one must resort to some kind of approximation or to suffer
the running time needed for a full stepping of the evolution equation.
The approximation available is on based on the PREM with simplifying
assumptions.  First, it assumes each earth shell to have a constant
density equal to the average over the neutrino track through that
shell.  Second, when calculating the average the density as a function
of radius is assumed to be linear.  


Implementation notes:
====================

The main interface are the functions in nuosc_prob.h.  There are 6
probability calculators covering the 3 matter effect possibilities
(vacuum, constant and PREM) times the 2 calculation methods (matrix
and step).  They are named like:

	nuosc_prob_[matter-type]_[calculation-type]()

With "matter-type" = "vacuum", "constant" or "prem" and
"calculation-type" = "matrix" or "step".

They take a ComplexVector (see matrix.h) holding the amplitude of the
initial neutrino, the oscilation parameters (bunded in OscParam, see
OscParam.h) as well as an energy and baseline.  Additionally, the
"constant" functions take a "density".  

The internal library units are uniform throughout:

	[dm^2]		= eV^2
	[angle]		= radians
	[length]	= cm
	[energy]	= eV
	[density]	= g/cm^3

See below for notes regarding units on the command line interfaces.

A note on anti-neutrinos: CP symetry is broken in that the matter
potential and the CP phase changes sign between neutrino and
anti-neutrino cases.  In the code this is handled inside by the
OscParam object via OscParam::set_antineutrino().  This unfortunately
mixes the idea of oscillation parameters (which are fundamental) with
the particular neutrino which is being evolved.  When asking OscParam
for a matrix via get_mixing_matrix() or get_matter_matrix() one will
have this -1 applied if using anti neutrinos.  However, the angle
returned via get_deltacp() will *NOT*.  When using these two matrices
for the evolution equation solution, one does not need to worry about
nu/antinu issues.  This is how nuosc_prob.cc functions behave.


Command line interface.
=======================

There exists two command line programs, "genprob" and the newer
"nuosc" (in main/) which give access to all these calculations.  The
former is deprecated in favor of the latter.

Simply running it w/out any args will produce a usage summary or run
with "nuosc -h" to get a detailed explanation of the command line
flags.

shell> nuosc -h
Option summary:
usage: nuosc [-n|--neutrino <initial neutrino number>]
             [-e|--energy <energy range description>]
             [-b|--baseline <baseline range description>]
             [-m|--mixing <mixing angle description>]
             [-s|--sol <delta m^2_solar (dm^2_21)>]
             [-a|--atm <delta m^2_atm (dm^2_31)>] [-d|--delta <CP phase>]
             [-D|--density <density description>]
             [-c|--calculation <calculation description] [-h|--help] 

Options details:
[-n|--neutrino <initial neutrino number>]
	Sets the initial neutrino e,mu,tau = 1,2,3; anti *= -1
	Default is 2 (mu)

[-e|--energy <energy range description>]
	Describe the energy range.  Can be one of:
		Linear energy range:      "lin:start,stop,step"
		Logarithmic energy range: "log:start,stop,step"
		A single energy value:    "one:energy"
	Linear energy units are in GeV,
	Logarithmic values give 10^start GeV - 10^stop GeV,
	Default is "one:1"

[-b|--baseline <baseline range description>]
	Describe the baseline range.  Can be one of:
		Linear disance range:       "lin:start,stop,step"
		Logarithmic distance range: "log:start,stop,step"
		A single distance value:    "one:distance"
		Zenith angle:               "zen:start,stop,step[,depth]"
		cos(Zenith angle):          "cos:start,stop,step[,depth]"
	Distance units are km, angle units are degrees.
	Logarithmic values give 10^start km - 10^stop km,
	The "depth" value for zenith related gives detector depth (def=0),
	Default is "one:1".

[-m|--mixing <mixing angle description>]
	Set the mixing angles description.  Can be one of:
		Angles:        "ang:theta_12,theta_23,theta_13"
		Sin^2(2theta): "sin:sin^2(2t_12),sin^2(2t_23),sin^2(2t_13)"
	Angle units are in degrees, default "sin:0.8,1.0,0.1"

[-s|--sol <delta m^2_solar (dm^2_21)>]
	Set the solar delta-m^2 (dm^2_21).
	Units are eV^2, default is 2.5e-3 eV^2

[-a|--atm <delta m^2_atm (dm^2_31)>]
	Set the atmospheric delta-m^2 (dm^2_31).
	Units are eV^2, default is 5.0e-5 eV^2

[-d|--delta <CP phase>]
	Set the CP phase angle.
	Units are degrees, default is 0 degrees

[-D|--density <density description>]
	Set the density description.  Can be one of:
		Constant density: "con:density"
		PREM density:     "prem"
		Lookup table:     "lut:filename"
	Density is in g/cc.
	The lookup text file has columns of bin centered position and density
	Default is "con:0.0"

[-c|--calculation <calculation description]
	Set the calculation description.  Can be one of:
		Matrix method: "matrix"
		Full stepping: "step"
	Default is "matrix"

[-h|--help]
	Print this message


The ouput is 

	<energy> <baseline> <P(nux->nue)> <P(nux->numu)> <P(nux->nutau)>

Notes:

 - The units are different ([angle]=degrees, [energy]=GeV,
 [distance]=km) than the underlying code uses and are given in the
 help message.

 - Although defaults are given for all options, one must specify at
 least one argument to actually do a calculation.

 - The faster "matrix" method only works on constant matter densities.
 If a calculation is done with varying densities, an density averaged
 along the baseline is used.  In the case of PREM, each major section
 (crust, mantle, core) is separately averaged.



ChangeLog
2007-04-03  Brett Viren  <bv@bnl.gov>

	* Convert build to autotools.

	* Only build "nuosc" for now.
	



main/Config.cc
#include "Config.h"
#include "util.h"
#include "options.h"

#include <cmath>
#include <iostream>
#include <vector>
#include <string>
#include <sstream>
using namespace std;

const char* optv[] = {
    "n:neutrino <initial neutrino number>",
    "e:energy <energy range description>",
    "b:baseline <baseline range description>",
    "m:mixing <mixing angle description>",
    "s:sol <delta m^2_solar (dm^2_21)>",
    "a:atm <delta m^2_atm (dm^2_31)>",
    "d:delta <CP phase>",
    "D:density <density description>",
    "c:calculation <calculation description",
    "h|help",
    0
};

// Describe each option
const char* desc[] = {
    // neutrino
    "\tSets the initial neutrino e,mu,tau = 1,2,3; anti *= -1\n"
    "\tDefault is 2 (mu)\n",

    // energy
    "\tDescribe the energy range.  Can be one of:\n"
    "\t\tLinear energy range:      \"lin:start,stop,step\"\n"
    "\t\tLogarithmic energy range: \"log:start,stop,step\"\n"
    "\t\tA single energy value:    \"one:energy\"\n"
    "\tLinear energy units are in GeV,\n"
    "\tLogarithmic values give 10^start GeV - 10^stop GeV,\n"
    "\tDefault is \"one:1\"\n",
    
    // baseline
    "\tDescribe the baseline range.  Can be one of:\n"
    "\t\tLinear disance range:       \"lin:start,stop,step\"\n"
    "\t\tLogarithmic distance range: \"log:start,stop,step\"\n"
    "\t\tA single distance value:    \"one:distance\"\n"
    "\t\tZenith angle:               \"zen:start,stop,step[,depth]\"\n"
    "\t\tcos(Zenith angle):          \"cos:start,stop,step[,depth]\"\n"
    "\tDistance units are km, angle units are degrees.\n"
    "\tLogarithmic values give 10^start km - 10^stop km,\n"
    "\tThe \"depth\" value for zenith related gives detector depth (def=0),\n"
    "\tDefault is \"one:1\".\n",

    // mixing
    "\tSet the mixing angles description.  Can be one of:\n"
    "\t\tAngles:        \"ang:theta_12,theta_23,theta_13\"\n"
    "\t\tSin^2(2theta): \"sin:sin^2(2t_12),sin^2(2t_23),sin^2(2t_13)\"\n"
    "\tAngle units are in degrees, default \"sin:0.8,1.0,0.1\"\n",

    // sol
    "\tSet the solar delta-m^2 (dm^2_21).\n"
    "\tUnits are eV^2, default is 2.5e-3 eV^2\n",

    // atm
    "\tSet the atmospheric delta-m^2 (dm^2_31).\n"
    "\tUnits are eV^2, default is 5.0e-5 eV^2\n",
    
    // delta
    "\tSet the CP phase angle.\n"
    "\tUnits are degrees, default is 0 degrees\n",

    // density
    "\tSet the density description.  Can be one of:\n"
    "\t\tConstant density: \"con:density\"\n"
    "\t\tPREM density:     \"prem\"\n"
    "\t\tLookup table:     \"lut:filename\"\n"
    "\tDensity is in g/cc.\n"
    "\tThe lookup text file has columns of bin centered position and density\n"
    "\tDefault is \"con:0.0\"\n",

    // calculation
    "\tSet the calculation description.  Can be one of:\n"
    "\t\tMatrix method: \"matrix\"\n"
    "\t\tFull stepping: \"step\"\n"
    "\tDefault is \"matrix\"\n",

    // help
    "\tPrint this message\n",

    0
};

void Config::usage(const char* message, bool full)
{
    cerr << "Option summary:" << endl;
    options->usage(cerr,"");

    
    if (full) {
	cerr << "\nOptions details:\n";
	for (int ind=0; optv[ind]; ++ind) {
	    string odesc(optv[ind]);
	    cerr << "[-" << odesc[0] << "|--" << odesc.substr(2) << "]\n";
	    cerr << desc[ind] << endl;
	}
    }

    if (message)
	cerr << "Error:\n\t" << message << endl;
    exit (1);
}

static double unfrob(double s22t) { return asin(sqrt(s22t))/2.0; }

static bool set_mixing(OscParam& op, string desc)
{
    vector<double> vd = strings2doubles(split((desc.substr(4)),','));
    if (vd.size() != 3) return false;
    
    if (desc.substr(0,3) == "ang") {
	op.set_theta12(vd[0]);
	op.set_theta23(vd[1]);
	op.set_theta13(vd[2]);
    }
    else if (desc.substr(0,3) == "sin") {
	op.set_theta12(unfrob(vd[0]));
	op.set_theta23(unfrob(vd[1]));
	op.set_theta13(unfrob(vd[2]));
    }
    else return false;
    return true;
}

Config::Config(int argc, const char** argv)
{
    // defaults
    nu_num = 2;
    energy_desc = "one:1";
    baseline_desc = "one:1";
    calc_desc = "matrix";
    density_desc = "con:0.0";

    options = new Options(*argv,optv);
    OptArgvIter optitr(argc-1,argv+1);
    const char* optarg;

    if(argc==1) usage("No args supplied");

    char optchar;
    while ((optchar = (*options)(optitr,optarg))) {
        switch (optchar) {
        case 'n':		// neutrino
	    if (!optarg) usage("No number given with -n");
	    nu_num = atoi(optarg);
	    op.set_antineutrino(nu_num < 0);
	    break;
        case 'e':		// energy
	    if (!optarg) usage("No energy description given with -e");
	    energy_desc = optarg;
	    break;
        case 'b':		// baseline
	    if (!optarg) usage("No baseline description given with -b");
	    baseline_desc = optarg;
	    break;
        case 'm':		// mixing 
	    if (!optarg) usage("No mixing description given with -m");
	    if (!set_mixing(op,optarg))
		usage("Failed to parse mixing description");
	    break;
        case 's':		// sol
	    if (!optarg) usage("No solar dm^2 given with -s");
	    op.set_dms21(atof(optarg));
	    break;
        case 'a':		// atm
	    if (!optarg) usage("No atmospheric dm^2 given with -a");
	    op.set_dms31(atof(optarg));
	    break;
        case 'd':		// delta
	    if (!optarg) usage("No CP phase angle given with -d");
	    op.set_deltacp(atof(optarg));
	    break;
        case 'D':		// density
	    if (!optarg) usage("No density description given with -D");
	    density_desc = optarg;
	    break;
        case 'c':		// calculation
	    if (!optarg) usage("No calculation description given with -c");
	    calc_desc = optarg;
	    break;
	case 'h':
	    usage(0,true);
	default:
	    usage("Unknown option given");
	    break;
	}
    }
}

Config::~Config() { delete options; options = 0; }

string Config::as_string() const
{
    stringstream ss;

    if (op.is_antineutrino()) ss << "anti-";
    switch(abs(nu_num)) {
    case 1: ss << "nu_e\n"; break;
    case 2: ss << "nu_mu\n"; break;
    case 3: ss << "nu_tau\n"; break;
    default: cerr << " UNKNOWN NEUTRINO\n"; break;
    }

    ss << "dms21   = " << op.get_dms21() << endl
       << "dms31   = " << op.get_dms31() << endl
       << "theta12 = " << op.get_theta12() << endl
       << "theta23 = " << op.get_theta23() << endl
       << "theta13 = " << op.get_theta13() << endl
       << "ss2t12  = " << op.get_ss2t12() << endl
       << "ss2t23  = " << op.get_ss2t23() << endl
       << "ss2t13  = " << op.get_ss2t13() << endl
       << "deltacp = " << op.get_deltacp() << endl
       << "calculation = (" << calc_desc << ")" << endl;
    ss << "baseline = " << baseline_desc << endl
       << "energy = " << energy_desc << endl
       << "density = " << density_desc << endl;

    return ss.str();
}
